common/flatpak-transaction.c

```diff
@@ -4081,6 +4114,325 @@ flatpak_transaction_normalize_ops (FlatpakTransaction *self)
+    }
+
+  unused_refs = flatpak_dir_list_unused_refs_with_options (priv->dir,
+                                                           NULL, /* arch */
+                                                           metadata_injection,
+                                                           to_be_excluded_strv,
+                                                           TRUE, /* filter_by_eol */
+                                                           cancellable, error);
+  if (unused_refs == NULL)
+    return FALSE;
+
+  GLNX_HASH_TABLE_FOREACH_IT (maybe_unused_runtimes, hashiter, const char *, runtime, const char *, remote)
+    {
+      if (!g_strv_contains ((const gchar * const *)unused_refs, runtime))
+        g_hash_table_iter_remove (&hashiter);
+    }
+
+  return TRUE;
+}
+
+static gboolean
+populate_maybe_unused_list (FlatpakTransaction  *self,
+                            GHashTable          *maybe_unused_runtimes,
+                            GCancellable        *cancellable,
+                            GError             **error)
+{
+  FlatpakTransactionPrivate *priv = flatpak_transaction_get_instance_private (self);
+  g_auto(GStrv) unused_refs = NULL;
+
+  unused_refs = flatpak_dir_list_unused_refs_with_options (priv->dir,
+                                                           NULL, /* arch */
+                                                           NULL, /* metadata_injection */
+                                                           NULL, /* refs_to_exclude */
+                                                           TRUE, /* filter_by_eol */
+                                                           cancellable, error);
+  if (unused_refs == NULL)
+    return FALSE;
+
+  for (char **iter = unused_refs; iter && *iter; iter++)
+    g_hash_table_replace (maybe_unused_runtimes, g_strdup (*iter), NULL);
+
+  return TRUE;
+}
+
+static gboolean
+add_uninstall_unused_ops (FlatpakTransaction  *self,
+                          GCancellable        *cancellable,
+                          GError             **error)
+{
+  FlatpakTransactionPrivate *priv = flatpak_transaction_get_instance_private (self);
+  g_autoptr(GHashTable) maybe_unused_runtimes = NULL;
+  g_autoptr(GHashTable) newly_used_runtimes = NULL;
+  g_autoptr(GHashTable) metadata_injection = NULL;
+  g_autoptr(GPtrArray) to_be_excluded = NULL;
+  g_autoptr(GPtrArray) run_after_ops = NULL;
+  GList *l, *next;
+  int i;
+
+  if (priv->disable_deps)
+    return TRUE;
+
+  /* This is the set of runtimes which are no longer needed by something in the
+   * transaction (an uninstall or update). The values are either the relevant
+   * remote name or %NULL. */
+  maybe_unused_runtimes = g_hash_table_new_full (g_str_hash, g_str_equal, g_free, g_free);
+
+  /* This is the set of runtimes being used by an install or update operation
+   * in the transaction. */
+  newly_used_runtimes = g_hash_table_new_full (g_str_hash, g_str_equal, g_free, NULL);
```

Comment:
Do we need newly_used_runtimes at all now? It seems to me that since we add all new metadatas as overrides it will never think that these new runtimes could be unused.